### ISSUE TITLE
Rule / Modules: Should they use a separate scope?

### ISSUE THREAD

**Issue by maggu2810** [eclipse/smarthome#3114]
Perhaps I am wrong, so please me correct me...

The different module handles add information to the global scope and this information could be used by e.g. Scripts.

E.g. some different modules add different stuff, e.g. state, newState, oldState, event, ...

All this is added to the global namespace...
Wouldn't it be better placed in a separate namespace?

So, stuff injected by handlers does not collide some time with stuff that is added by the rule themselves?

So e.g. ctx.newState, ctx.oldState, ...
Or do I miss there something?

**Comment by danchom** [comment]
@maggu2810 What context do you mean? In the automation engine context only outputs from calculated modules are put in it. They are in form moduleID.outputId = value. Also the input values of connected inputs are  also put in, in form of inputId=value. Other modules can't change automation context as we have discussed that in #3021.

**Comment by maggu2810** [comment]
I talk about the context the script is using (not really about the context that is used in the automation source code).

Let's assume I am an user of the Paper UI and would like to create a new Rule:

* create a new rule
* add trigger `an item state changes`
* add action `execute a given script` (JavaScript)
* there is now `newState`, `oldState` and `event`

I can add code like:

```JavaScript
print(newState);
print(oldState);
```

So, that information is in the global JavaScript execution namespace.
So `oldState` and `newState` are already set by the engine and will potentially collide with variables in my JavaScript code.

Is this wrong?

**Comment by danchom** [comment]
@kaikreuzer may be you have to answer here. I'm not in deep in the scripting module handler implementation.

**Comment by maggu2810** [comment]
ping...

**Comment by kaikreuzer** [comment]
> So oldState and newState are already set by the engine and will potentially collide with variables in my JavaScript code.

Yes - but the JavaScript code should be specific for this individual rule, so it should be possible to avoid such collisions.
Note that the old DSL rule engine also has preset keywords that depend on the used trigger ("receivedCommand", "previousState", etc.). so this feature here in the new rule engine is in line with that.
